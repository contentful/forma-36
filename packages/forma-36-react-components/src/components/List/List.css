.List {
  margin: 0;
}

.List--unordered {
  list-style-type: disc;
}
