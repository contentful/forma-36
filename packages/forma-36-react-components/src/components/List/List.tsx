import React from 'react';
import cn from 'classnames';

import styles from './List.css';

type ListType = 'ul' | 'ol';

export interface ListProps {
  element?: ListType;
  className?: string;
  children: React.ReactNode;
  style?: React.CSSProperties;
  testId?: string;
}

export function List({
  element: Tag = 'ul',
  className,
  children,
  testId = 'cf-ui-list',
  ...otherProps
}: ListProps): React.ReactElement {
  const classNames = cn(
    styles['List'],
    { [styles['List--unordered']]: Tag === 'ul' },
    className,
  );

  return (
    <Tag {...otherProps} className={classNames} data-test-id={testId}>
      {children}
    </Tag>
  );
}

export default List;
