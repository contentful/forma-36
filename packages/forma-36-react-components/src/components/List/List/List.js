import React from 'react';
import cn from 'classnames';
import PropTypes from 'prop-types';
import styles from './List.css';

export class List extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    children: PropTypes.node.isRequired,
    testId: PropTypes.string,
  };

  static defaultProps = {
    extraClassNames: undefined,
    testId: 'cf-ui-list',
  };

  render() {
    const { extraClassNames, children, testId, ...otherProps } = this.props;

    const classNames = cn(styles.List, extraClassNames);

    return (
      <ul className={classNames} data-test-id={testId} {...otherProps}>
        {children}
      </ul>
    );
  }
}

export default List;
