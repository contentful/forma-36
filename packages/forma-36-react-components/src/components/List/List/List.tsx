import React, { Component } from 'react';
import cn from 'classnames';
import styles from './List.css';

export type ListProps = {
  className?: string;
  children: React.ReactNode;
  style?: React.CSSProperties;
  testId?: string;
} & typeof defaultProps;

const defaultProps = {
  testId: 'cf-ui-list',
};

export class List extends Component<ListProps> {
  static defaultProps = defaultProps;

  render() {
    const { className, children, testId, ...otherProps } = this.props;

    const classNames = cn(styles['List'], className);

    return (
      <ul {...otherProps} className={classNames} data-test-id={testId}>
        {children}
      </ul>
    );
  }
}

export default List;
