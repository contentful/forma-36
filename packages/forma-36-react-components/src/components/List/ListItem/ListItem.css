@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';

.ListItem {
  font-size: var(--font-size-m);
  font-family: var(--font-stack-primary);
  line-height: var(--line-height-default);
}

.ListItem--nested-list {
  list-style: none;
}
