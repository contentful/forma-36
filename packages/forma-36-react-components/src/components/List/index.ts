export * from './List';
export { default } from './List';
