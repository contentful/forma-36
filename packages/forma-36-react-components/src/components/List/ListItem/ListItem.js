import React from 'react';
import cn from 'classnames';
import PropTypes from 'prop-types';
import styles from './ListItem.css';
import List from '../List/List';

export class ListItem extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    children: PropTypes.node.isRequired,
    testId: PropTypes.string,
  };

  static defaultProps = {
    extraClassNames: undefined,
    testId: 'cf-ui-list-item',
  };

  render() {
    const { extraClassNames, children, testId, ...otherProps } = this.props;

    const classNames = cn(styles.ListItem, extraClassNames, {
      [styles['ListItem--nested-list']]: this.props.children.type === List,
    });

    return (
      <li className={classNames} data-test-id={testId} {...otherProps}>
        {children}
      </li>
    );
  }
}

export default ListItem;
