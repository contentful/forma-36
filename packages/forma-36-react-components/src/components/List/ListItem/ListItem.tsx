import React, { Component } from 'react';
import cn from 'classnames';

import styles from './ListItem.css';
import List from '../List/List';

export interface ListItemProps {
  extraClassNames?: string;
  children: React.ReactNode;
  testId?: string;
}

export class ListItem extends Component<ListItemProps> {
  static defaultProps = {
    testId: 'cf-ui-list-item',
  };

  render() {
    const { extraClassNames, children, testId, ...otherProps } = this.props;

    const classNames = cn(styles['ListItem'], extraClassNames, {
      [styles['ListItem--nested-list']]:
        (this.props.children as any).type === List,
    });

    return (
      <li className={classNames} data-test-id={testId} {...otherProps}>
        {children}
      </li>
    );
  }
}

export default ListItem;
