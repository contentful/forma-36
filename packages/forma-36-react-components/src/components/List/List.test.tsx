import React from 'react';
import { render } from '@testing-library/react';

import axe from '../../utils/axeHelper';
import List from './List';
import ListItem from './ListItem';

it('renders the component', () => {
  const { container } = render(
    <List>
      <ListItem>Item</ListItem>
    </List>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders as ordered list', () => {
  const { container } = render(
    <List element="ol">
      <ListItem>Item</ListItem>
    </List>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const { container } = render(
    <List className="my-extra-class">
      <ListItem>Item</ListItem>
    </List>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const { container } = render(
    <List>
      <ListItem>Item</ListItem>
    </List>,
  );
  const results = await axe(container);

  expect(results).toHaveNoViolations();
});
