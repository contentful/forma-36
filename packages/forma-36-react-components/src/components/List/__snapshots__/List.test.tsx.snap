// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as ordered list 1`] = `
<ol
  class="List"
  data-test-id="cf-ui-list"
>
  <li
    class="ListItem"
    data-test-id="cf-ui-list-item"
  >
    Item
  </li>
</ol>
`;

exports[`renders the component 1`] = `
<ul
  class="List List--unordered"
  data-test-id="cf-ui-list"
>
  <li
    class="ListItem"
    data-test-id="cf-ui-list-item"
  >
    Item
  </li>
</ul>
`;

exports[`renders the component with an additional class name 1`] = `
<ul
  class="List List--unordered my-extra-class"
  data-test-id="cf-ui-list"
>
  <li
    class="ListItem"
    data-test-id="cf-ui-list-item"
  >
    Item
  </li>
</ul>
`;
